File: python-apiops-agentlab/app/schemas/diagnosis_report.py
"""Strict Pydantic v2 view for the shared DiagnosisReport contract."""

from __future__ import annotations

from typing import Annotated, Any, Literal, TypeAlias

from pydantic import BaseModel, ConfigDict, Field, StrictBool, StrictInt, StrictStr

FailureType: TypeAlias = Literal[
    "NONE",
    "ASSERTION_MISMATCH",
    "ASSERTION_EVALUATION_ERROR",
    "HTTP_STATUS_ERROR",
    "TIMEOUT",
    "NETWORK_ERROR",
    "REQUEST_BUILD_ERROR",
    "INVALID_TARGET_URI",
    "DNS_ERROR",
    "CONNECT_ERROR",
    "TLS_ERROR",
    "IO_ERROR",
    "SCHEMA_INVALID",
    "BUSINESS_ERROR",
    "TOOL_ERROR",
    "SYSTEM_ERROR",
    "UNKNOWN",
    "ASSERTION_FAILED",
    "AUTH_ERROR",
    "VALIDATION_ERROR",
    "DATABASE_CONSTRAINT_ERROR",
    "UPSTREAM_SERVICE_ERROR",
]

Confidence: TypeAlias = Literal["LOW", "MEDIUM", "HIGH"]
NonEmptyString: TypeAlias = Annotated[StrictStr, Field(min_length=1)]


class _DiagnosisModel(BaseModel):
    model_config = ConfigDict(
        extra="forbid",
        populate_by_name=False,
        strict=True,
    )

    def model_dump(self, **kwargs: Any) -> dict[str, Any]:
        kwargs.setdefault("by_alias", True)
        return super().model_dump(**kwargs)


class EvidenceRef(_DiagnosisModel):
    """Structural citation identity; contextual existence is validated elsewhere."""

    item_id: NonEmptyString = Field(alias="itemId")


class RootCauseHypothesis(_DiagnosisModel):
    statement: NonEmptyString
    confidence: Confidence
    evidence_refs: list[EvidenceRef] = Field(alias="evidenceRefs", min_length=1)


class DiagnosisReport(_DiagnosisModel):
    """Typed semantic inference result, separate from Java execution facts.

    The wire contract keeps the historic ``failureType`` name.  Within Python
    that value is the Agent's semantic diagnosis; the observed runner failure
    remains available only from the authoritative ``TestReport``.
    """

    schema_version: Literal["0.1.0"] = Field(alias="schemaVersion")
    report_id: NonEmptyString = Field(alias="reportId")
    agent_run_id: NonEmptyString = Field(alias="agentRunId")
    project_id: StrictInt = Field(alias="projectId", ge=1)
    run_id: StrictInt = Field(alias="runId", ge=1)
    failure_type: FailureType = Field(alias="failureType")
    summary: NonEmptyString
    root_cause_hypotheses: list[RootCauseHypothesis] = Field(alias="rootCauseHypotheses")
    sufficient_evidence: StrictBool = Field(alias="sufficientEvidence")
    limitations: list[NonEmptyString]
    recommended_checks: list[NonEmptyString] = Field(alias="recommendedChecks")
    trace_id: NonEmptyString = Field(alias="traceId")

    @property
    def semantic_diagnosis(self) -> FailureType:
        """Return the Agent-authored class without implying a Java observation."""

        return self.failure_type

File: java-apiops-platform/apiops-agent/src/main/java/com/apiops/agent/diagnosis/DiagnosisReport.java
package com.apiops.agent.diagnosis;

import java.util.List;
import java.util.Objects;

/** Model reasoning output; it is never written into execution facts. */
public record DiagnosisReport(
        String schemaVersion,
        String reportId,
        String agentRunId,
        long projectId,
        long runId,
        String failureType,
        String summary,
        List<RootCauseHypothesis> rootCauseHypotheses,
        boolean sufficientEvidence,
        List<String> limitations,
        List<String> recommendedChecks,
        String traceId
) {

    public DiagnosisReport {
        schemaVersion = requireText(schemaVersion, "schemaVersion");
        reportId = requireText(reportId, "reportId");
        agentRunId = requireText(agentRunId, "agentRunId");
        failureType = requireText(failureType, "failureType");
        summary = requireText(summary, "summary");
        traceId = requireText(traceId, "traceId");
        if (projectId <= 0 || runId <= 0) {
            throw new IllegalArgumentException("projectId and runId must be positive");
        }
        rootCauseHypotheses = List.copyOf(Objects.requireNonNull(
                rootCauseHypotheses, "rootCauseHypotheses must not be null"));
        limitations = textList(limitations, "limitations");
        recommendedChecks = textList(recommendedChecks, "recommendedChecks");
    }

    private static List<String> textList(List<String> values, String field) {
        Objects.requireNonNull(values, field + " must not be null");
        List<String> copy = List.copyOf(values);
        if (copy.stream().anyMatch(value -> value == null || value.isBlank())) {
            throw new IllegalArgumentException(field + " must contain non-blank text");
        }
        return copy;
    }

    private static String requireText(String value, String field) {
        Objects.requireNonNull(value, field + " must not be null");
        if (value.isBlank()) {
            throw new IllegalArgumentException(field + " must not be blank");
        }
        return value;
    }
}
